safenode/src/storage/spends.rs

```diff
@@ -79,7 +72,26 @@ impl SpendStorage {
     /// If a double spend attempt is detected, a `DoubleSpendAttempt` error
     /// will be returned including all the `SignedSpends`, for
     /// broadcasting to the other nodes.
-    pub(super) async fn try_add(&self, signed_spend: &SignedSpend) -> Result<()> {
+    pub(crate) async fn try_add(&self, signed_spend: &SignedSpend) -> Result<()> {
+        self.validate(signed_spend).await?;
+
+        let size_of_new = std::mem::size_of_val(signed_spend);
+        let address = dbc_address(signed_spend.dbc_id());
+
+        let mut valid_spends = self.valid_spends.write().await;
+
+        let replaced = valid_spends.insert(address, signed_spend.clone());
+        if replaced.is_none() {
+            self.valid_spends_cache_size.increase(size_of_new);
+        }
+
```

Comment:
Yes, that is right.

There is a call to `self.validate` before, which checks for double spends. 

Just needs to update the fn signature to take `&mut self`, to avoid having something added in between that check, and this insert.

Good catch!